### PR TITLE
[Routing] don't process classes with the directory loader

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25225
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php [hunk 1]
```diff
@@ -76,6 +76,15 @@ function (\SplFileInfo $current) {
      */
     public function supports($resource, $type = null)
     {
+        if ($type && 'annotation' !== $type) {
+            return false;
+        }
+
+        // classes are handled by the AnnotationClassLoader
+        if (preg_match('/^(?:\\\\?[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)+$/', $resource)) {
```
Comment (dmaicher) [inline src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php]: wondering why it worked on 3.3 without this? Was the order of the loaders different?

Comment (xabbuh) [reply src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php]: Actually, I think we really need to make the change in 3.3. The observed behaviour seems to be caused by #25113 which just was not part of a patch release of 3.3 yet.

Comment (dmaicher) [reply src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php]: Ah ok I see :wink: 

Can't we do a `class_exists` call to see if `$resource` is a class instead?

**Comment by nicolas-grekas** [comment]
What about changing the priority of the loaders instead?

**Comment by xabbuh** [comment]
@nicolas-grekas I think the root issue is that paths are different when `supports()` and `load()` are called. The reason is that the `FileLoader` resolves paths relative to the current dir after `supports()` but before `load()` is called.

**Comment by xabbuh** [comment]
@nicolas-grekas I have pushed a second commit which could be the solution. Basically, it reverts the changes done in #25113 and does the resource resolving inside the `FileLoader` if resolving the loader fails. This should ensure that the `supports()` method receives the same resource as the `load()` method later.

File: src/Symfony/Component/Config/Loader/FileLoader.php [hunk 2]
```diff
@@ -130,7 +130,18 @@ protected function glob($pattern, $recursive, &$resource = null, $ignoreErrors =
...
+                    $resource = $loader->getLocator()->locate($resource, $this->currentDir, false);
+                }
+            } catch (\Exception $e) {
+                // as a last resort try to find a loader after locating the resource
+                $resource = $this->getLocator()->locate($resource, $this->currentDir, false);
+                $resource = is_array($resource) ? $resource[0] : $resource;
+                $loader = $this->resolve($resource, $type);
+            }
 
             if ($loader instanceof self && null !== $this->currentDir) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Loader/FileLoader.php]: Duplicate?

Comment (xabbuh) [reply src/Symfony/Component/Config/Loader/FileLoader.php]: removed from the `try` block

**Comment by nicolas-grekas** [comment]
Red :)

File: src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php [hunk 3]
```diff
@@ -76,18 +74,16 @@ function (\SplFileInfo $current) {
      */
     public function supports($resource, $type = null)
     {
-        if ('annotation' === $type) {
-            return true;
-        }
-
-        if ($type || !is_string($resource)) {
+        if (!is_string($resource)) {
```
Comment (dmaicher) [inline src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php]: can be 

```php
if (!is_string($resource) || $type && 'annotation' !== $type) {
```

and then no need to change the try block below?

**Comment by nicolas-grekas** [comment]
I think we should go down this road: this changes the locator to be fetched from "$loader" vs "$this" in some cases, and this is too risky for the base FileLoader IMHO.
I would revert to the first version of the patch, which excluded FQCN.
to make it a bit more strict and less likely to break, I would also suggest to exclude only if the resource contains a `\` and to add a class_exists check in AnnotationClassLoader::supports.

File: src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php [hunk 4]
```diff
@@ -76,6 +76,15 @@ function (\SplFileInfo $current) {
      */
     public function supports($resource, $type = null)
     {
+        if ($type && 'annotation' !== $type) {
+            return false;
+        }
+
+        // classes are handled by the AnnotationClassLoader
+        if (strpos($resource, '\\') || class_exists($resource)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php]: `false !== strpos` to accurately mirror what's done in AnnotationClassLoader?

Comment (xabbuh) [reply src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php]: of course

File: src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php [hunk 5]
```diff
@@ -76,6 +76,15 @@ function (\SplFileInfo $current) {
      */
     public function supports($resource, $type = null)
     {
+        if ($type && 'annotation' !== $type) {
+            return false;
+        }
+
+        // classes are handled by the AnnotationClassLoader
+        if (false !== strpos($resource, '\\') || class_exists($resource)) {
```
Comment (stof) [inline src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php]: that's not good. Importing `%kernel.project_dir%/src/Controller` will have backslashes in it on Windows AFAICT.

Comment (nicolas-grekas) [reply src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php]: so we should add the regexp check also

File: src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php [hunk 6]
```diff
@@ -76,6 +76,15 @@ function (\SplFileInfo $current) {
      */
     public function supports($resource, $type = null)
     {
+        if ($type && 'annotation' !== $type) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php]: replaces the checks below, so they could be removed, but misses the `is_string($resource)` one

**Comment by nicolas-grekas** [comment]
there is something strange:
in FrameworkExtension
- AnnotatedRouteControllerLoader
- AnnotationDirectoryLoader
- AnnotationFileLoader

they're registered with priority -10, in this order
I'd expect the issue to not exist
because AnnotatedRouteControllerLoader (=AnnotationClasLoader) is first

related to "sort" which is not "stable" anymore since 7.0? maybe that's a better place for a fix, making this sorting stable?

**Comment by stof** [comment]
@nicolas-grekas could it be related to SensioFrameworkExtraBundle loaders having a different priority ?

**Comment by xabbuh** [comment]
The order in FrameworkExtraBundle is different:

* `AnnotationDirectoryLoader`
* `AnnotationFileLoader`
* `AnnotatedRouteControllerLoader`

**Comment by xabbuh** [comment]
So, should we maybe just change the order of the services in SensioFrameworkExtraBundle instead?

**Comment by nicolas-grekas** [comment]
we should for sure, for consistency at least!

**Comment by xabbuh** [comment]
sensiolabs/SensioFrameworkExtraBundle#537

**Pull request closed without merging**